hubble/cmd/common/conn/conn.go

```diff
@@ -77,3 +83,125 @@ func New(ctx context.Context, target string, dialTimeout time.Duration) (*grpc.C
+	// and launch a kubectl port-forward process
+	if vp.GetBool(config.KeyAutoPortForward) {
+		// inspired by: cilium-cli/hubble/relay.go:RelayPortForwardCommand()
+		// TODO: should we try to consolidate with cilium-cli ?
+		// TODO: second iteration: replace with native port-forward using k8s client (ProxyTCP())
+		k8sContextName := vp.GetString(config.KeyK8sContextName)
+		k8sKubeconfig := vp.GetString(config.KeyK8sKubeconfig)
+		k8sNamespace := vp.GetString(config.KeyK8sNamespace)
+		localPort := vp.GetUint16(config.KeyPortForward)
+
+		k8sClient, err := k8s.NewClient(k8sContextName, k8sKubeconfig)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		logger.Logger.Debug("Querying kubernetes API for hubble-relay service")
+		relaySvc, err := k8sClient.GetService(ctx, k8sNamespace, "hubble-relay", metav1.GetOptions{})
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// TODO: re-using grpc dial timeout, should we hardcode this or have another flag?
+		portForwardAddr, portForwardCleanup, err := kubectlPortForward(ctx, k8sNamespace, k8sContextName, dialTimeout, int(localPort), int(relaySvc.Spec.Ports[0].Port))
+		if err != nil {
+			return nil, nil, err
+		}
+		server = portForwardAddr
+		cleanupFns = append(cleanupFns, portForwardCleanup)
+	}
+
+	conn, err = New(ctx, server, dialTimeout)
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanupFns = append(cleanupFns, conn.Close)
+
+	cleanup = func() error {
+		var err error
+		for i := len(cleanupFns) - 1; i >= 0; i-- {
+			err = errors.Join(err, cleanupFns[i]())
+		}
+		return err
+	}
+	return conn, cleanup, nil
+}
+
+func kubectlPortForward(ctx context.Context, namespace, contextName string, dialTimeout time.Duration, local, remote int) (addr string, cleanup func() error, err error) {
+	localAddr := fmt.Sprintf("127.0.0.1:%d", local)
+
+	logger.Logger.Debug("Checking if port is already opened for port forwarding", "addr", localAddr)
+	if err := doTCPProbe(localAddr, dialTimeout); err == nil {
+		return "", nil, fmt.Errorf("cannot setup port-forwarding: already listening: %s", localAddr)
+	}
+
+	args := []string{
+		"port-forward",
+		"-n", namespace,
+		"svc/hubble-relay",
+		"--address", "127.0.0.1",
+		fmt.Sprintf("%d:%d", local, remote)}
+	if contextName != "" {
+		args = append([]string{"--context", contextName}, args...)
+	}
+
+	logger.Logger.Debug("Launching kubectl command", "args", args)
+	// TODO: is there a cleaner/more succint way to handle this?
+	ctx, cancel := context.WithCancel(ctx)
+	cmd := exec.CommandContext(ctx, "kubectl", args...)
+	cleanup = func() error {
+		cancel()
+		return cmd.Wait()
+	}
+	defer func() {
+		if err != nil {
+			cleanup()
+		}
+	}()
+	if err := cmd.Start(); err != nil {
+		return "", nil, err
+	}
+
+	maxRetries := 10
+	retryInterval := time.Second
+	retries := 0
+	for {
+		if retries > maxRetries {
+			return "", nil, fmt.Errorf("cannot setup port-forwarding: max retry reached (%d)", maxRetries)
+		}
+		time.Sleep(retryInterval)
```

Comment:
Probably good idea to use a `select` here with `time.After` and the `context.Done()` chan

Reply:
Good call out, done :)